sonic-xcvrd/xcvrd/xcvrd.py

```diff
@@ -2351,36 +2400,54 @@ def run(self):
         port_mapping_data, retry_eeprom_set = self.init()
 
         # Start the CMIS manager
-        cmis_manager = CmisManagerTask(self.namespaces, port_mapping_data, self.skip_cmis_mgr)
-        cmis_manager.task_run()
+        cmis_manager = CmisManagerTask(self.namespaces, port_mapping_data, self.stop_event, self.skip_cmis_mgr)
+        cmis_manager.start()
+        self.threads.append(cmis_manager)
 
         # Start the dom sensor info update thread
-        dom_info_update = DomInfoUpdateTask(self.namespaces, port_mapping_data)
-        dom_info_update.task_run()
+        dom_info_update = DomInfoUpdateTask(self.namespaces, port_mapping_data, self.stop_event)
+        dom_info_update.start()
+        self.threads.append(dom_info_update)
 
-        # Start the sfp state info update process
-        sfp_state_update = SfpStateUpdateTask(self.namespaces, port_mapping_data, retry_eeprom_set)
-        sfp_state_update.task_run(self.sfp_error_event)
-
-        # Start the Y-cable state info update process if Y cable presence established
+        # Start the sfp state info update thread
+        sfp_state_update = SfpStateUpdateTask(self.namespaces, port_mapping_data, retry_eeprom_set, self.stop_event, self.sfp_error_event)
+        sfp_state_update.start()
+        self.threads.append(sfp_state_update)
 
         # Start main loop
-        self.log_info("Start daemon main loop")
+        self.log_info("Start daemon main loop with thread count {}".format(len(self.threads)))
 
         self.stop_event.wait()
 
         self.log_info("Stop daemon main loop")
 
+        generate_sigkill = False
+        # check all threads are alive
+        for thread in self.threads:
+            if thread.is_alive() is False:
+                try:
+                    thread.join()
+                except Exception as e:
+                    self.log_error("Xcvrd: exception found at child thread {} due to {}".format(thread.getName(), repr(e)))
+                    generate_sigkill = True
+
+        if generate_sigkill is True:
+            self.log_error("Exiting main loop as child thread raised exception!")
+            os.kill(os.getpid(), signal.SIGKILL)
+
         # Stop the CMIS manager
         if cmis_manager is not None:
-           cmis_manager.task_stop()
+            if cmis_manager.is_alive():
+                cmis_manager.join()
 
         # Stop the dom sensor info update thread
-        dom_info_update.task_stop()
-
-        # Stop the sfp state info update process
-        sfp_state_update.task_stop()
+        if dom_info_update.is_alive():
+            dom_info_update.join()
 
+        # Stop the sfp state info update thread
+        if sfp_state_update.is_alive():
+            sfp_state_update.raise_exception()
```

Comment:
Will it be good to join(1) with timeout and raise exception if timeout expires and the thread is still alive?

Reply:
I feel it's better to raise exception right away since based on my current analysis on Arista testbed, it takes about ~5s for the vendor api to come out of sleep. This timeout I assume could be dynamic based on the transceivers present on the device so I think having 1s delay will not give us a much advantage for join function with 1s timeout.